File: kotlin/src/main/kotlin/aether/dtn/ReplicationStrategy.kt
// SPDX-License-Identifier: MIT

package aether.dtn

import aether.models.DtnBundle
import aether.models.PeerInfo

/**
 * Decides which connected peers should receive a copy of a bundle on the next
 * replication pass. Default GeohashEpidemicStrategy matches the C# reference.
 */
interface ReplicationStrategy {
    fun selectTargets(
        bundle: DtnBundle,
        peers: List<PeerInfo>,
        localGeohash: String?
    ): List<String>
}

/**
 * Default geohash-aware epidemic strategy.
 *
 * SOS bundles fan out to every eligible carrier up to the copy cap. Normal
 * bundles prefer peers whose geohash shares a longer prefix with the recipient's
 * last-known geohash than the local node — i.e. peers at least as close to the
 * recipient. Ties broken by reliability score.
 */
class GeohashEpidemicStrategy : ReplicationStrategy {
    override fun selectTargets(
        bundle: DtnBundle,
        peers: List<PeerInfo>,
        localGeohash: String?
    ): List<String> {
        val slots = bundle.maxCopies - bundle.copyCount
        if (slots <= 0) return emptyList()

        val eligible = peers.filter {
            it.uhid.isNotEmpty()
                && it.uhid != bundle.senderUhid
                && it.capabilities.dtnCarrier
        }
        if (eligible.isEmpty()) return emptyList()

        // SOS priority = 3
        if (bundle.priority == 3) {
            return eligible.take(slots).map { it.uhid }
        }

        if (!bundle.recipientLastGeohash.isNullOrEmpty()) {
            val recipient = bundle.recipientLastGeohash
            val localProx = sharedPrefix(localGeohash, recipient)
            return eligible
                .map { p -> Triple(sharedPrefix(p.geohash, recipient), p.reliabilityScore, p) }
                .filter { (prox, _, _) -> prox >= localProx }
                .sortedWith(compareByDescending<Triple<Int, Int, PeerInfo>> { it.first }
                    .thenByDescending { it.second })
                .take(slots)
                .map { it.third.uhid }
        }

        return eligible
            .sortedByDescending { it.reliabilityScore }
            .take(slots)
            .map { it.uhid }
    }

    private fun sharedPrefix(a: String?, b: String): Int {
        if (a.isNullOrEmpty() || b.isEmpty()) return 0
        val n = minOf(a.length, b.length)
        var i = 0
        while (i < n && a[i] == b[i]) i++
        return i
    }
}

File: swift/Sources/AetherNetProtocol/DTN/ReplicationStrategy.swift
// SPDX-License-Identifier: MIT

import Foundation

/// Decides which connected peers should receive a copy of a bundle on the next replication pass.
public protocol ReplicationStrategy: Sendable {
    func selectTargets(bundle: DtnBundle, peers: [PeerInfo], localGeohash: String?) -> [String]
}

/// Default geohash-aware epidemic strategy. SOS bundles fan out to every eligible carrier
/// up to the copy cap; normal bundles prefer peers whose geohash shares a longer prefix
/// with the recipient than the local node. Ties broken by reliability score.
public struct GeohashEpidemicStrategy: ReplicationStrategy {
    public init() {}

    public func selectTargets(bundle: DtnBundle, peers: [PeerInfo], localGeohash: String?) -> [String] {
        let slots = Int(bundle.maxCopies - bundle.copyCount)
        if slots <= 0 { return [] }

        let dtnFlag = NodeCapabilityBits.dtnCarrier
        let eligible = peers.filter { p in
            !p.uhid.isEmpty
                && p.uhid != bundle.senderUhid
                && !p.isBlocked
                && (p.capabilities & dtnFlag) == dtnFlag
        }
        if eligible.isEmpty { return [] }

        if bundle.priority == BundlePriority.sos.rawValue {
            return eligible.prefix(slots).map { $0.uhid }
        }

        if let recipient = bundle.recipientLastGeohash, !recipient.isEmpty {
            let localProx = sharedPrefix(localGeohash, recipient)
            let ranked = eligible
                .map { (peer: $0, prox: sharedPrefix($0.geohash, recipient)) }
                .filter { $0.prox >= localProx }
                .sorted {
                    if $0.prox != $1.prox { return $0.prox > $1.prox }
                    return $0.peer.reliabilityScore > $1.peer.reliabilityScore
                }
            return ranked.prefix(slots).map { $0.peer.uhid }
        }

        return eligible
            .sorted { $0.reliabilityScore > $1.reliabilityScore }
            .prefix(slots)
            .map { $0.uhid }
    }

    private func sharedPrefix(_ a: String?, _ b: String) -> Int {
        guard let a = a, !a.isEmpty, !b.isEmpty else { return 0 }
        let n = min(a.count, b.count)
        var i = 0
        let aChars = Array(a)
        let bChars = Array(b)
        while i < n && aChars[i] == bChars[i] { i += 1 }
        return i
    }
}
